core/src/server/access_control/origin.rs

```diff
@@ -0,0 +1,359 @@
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Origin filtering functions
+
+use std::{fmt, ops};
+
+use crate::server::access_control::host::{Host, Port};
+use crate::server::access_control::matcher::{Matcher, Pattern};
+use crate::Error;
+
+/// Origin Protocol
+#[derive(Clone, Hash, Debug, PartialEq, Eq)]
+pub enum OriginProtocol {
+	/// Http protocol
+	Http,
+	/// Https protocol
+	Https,
+	/// Custom protocol
+	Custom(String),
+}
+
+/// Request Origin
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct Origin {
+	protocol: OriginProtocol,
+	host: Host,
+	host_with_proto: String,
+	matcher: Matcher,
+}
+
+impl<T: AsRef<str>> From<T> for Origin {
+	fn from(origin: T) -> Self {
+		Origin::parse(origin.as_ref())
+	}
+}
+
+impl Origin {
+	fn with_host(protocol: OriginProtocol, host: Host) -> Self {
+		let host_with_proto = Self::host_with_proto(&protocol, &host);
+		let matcher = Matcher::new(&host_with_proto);
+
+		Origin { protocol, host, host_with_proto, matcher }
+	}
+
+	/// Creates new origin given protocol, hostname and port parts.
+	/// Pre-processes input data if necessary.
+	pub fn new<T: Into<Port>>(protocol: OriginProtocol, host: &str, port: T) -> Self {
+		Self::with_host(protocol, Host::new(host, port))
+	}
+
+	/// Attempts to parse given string as a `Origin`.
+	/// NOTE: This method always succeeds and falls back to sensible defaults.
+	pub fn parse(origin: &str) -> Self {
+		let mut parts = origin.split("://");
+		let proto = parts.next().expect("split always returns non-empty iterator.");
+		let hostname = parts.next();
+
+		let (proto, hostname) = match hostname {
+			None => (None, proto),
+			Some(hostname) => (Some(proto), hostname),
+		};
+
+		let proto = proto.map(str::to_lowercase);
+		let hostname = Host::parse(hostname);
+
+		let protocol = match proto {
+			None => OriginProtocol::Http,
+			Some(ref p) if p == "http" => OriginProtocol::Http,
+			Some(ref p) if p == "https" => OriginProtocol::Https,
+			Some(other) => OriginProtocol::Custom(other),
+		};
+
+		Origin::with_host(protocol, hostname)
+	}
+
+	fn host_with_proto(protocol: &OriginProtocol, host: &Host) -> String {
+		format!(
+			"{}://{}",
+			match *protocol {
+				OriginProtocol::Http => "http",
+				OriginProtocol::Https => "https",
+				OriginProtocol::Custom(ref protocol) => protocol,
+			},
+			&**host,
+		)
+	}
+}
+
+impl Pattern for Origin {
+	fn matches<'a, T: AsRef<str> + PartialEq<&'a str>>(&self, other: T) -> bool {
+		self.matcher.matches(other)
+	}
+}
+
+impl ops::Deref for Origin {
+	type Target = str;
+	fn deref(&self) -> &Self::Target {
+		&self.host_with_proto
+	}
+}
+
+/// Origin type allowed to access.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum OriginType {
+	/// Specific origin.
+	Origin(Origin),
+	/// Null origin (file:///, sandboxed iframe)
+	Null,
+	/// Any non-null origin.
```

Comment:
I don't really understand why this was re-named to `AnyNonNull` this naming is not clear to me. I still like `Any` or `Wildcard` better.

I guess the comment is from the old code but I suggest that we fix it because it's doesn't say anything useful at all.

> For requests without credentials, the literal value "*" can be specified as a wildcard; the value tells browsers to allow requesting code from any origin to access the resource

So I suggest stating `Allow all origins` or something like that

Reply:
Done! Picked `Wildcard`, as it seems a bit more expressive, thanks!